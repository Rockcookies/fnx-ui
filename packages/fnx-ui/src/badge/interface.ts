import { ReactNode, HTMLAttributes } from 'react';

export type BadgeProps = BadgeComponentProps &
	Omit<HTMLAttributes<HTMLSpanElement>, 'color'>;

export interface BadgeComponentProps {
	count?: ReactNode;
	showZero?: boolean;
	dot?: boolean;
	max?: string | number;
	color?: string;
	offset?: [number | string, number | string];
}
