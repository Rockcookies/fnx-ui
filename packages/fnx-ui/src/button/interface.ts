import { ButtonHTMLAttributes, ReactNode } from 'react';

export type ButtonType =
	| 'default'
	| 'primary'
	| 'success'
	| 'warning'
	| 'danger';

export type ButtonSize = 'lg' | 'md' | 'sm' | 'xs';

export type ButtonShape = 'rect' | 'radius' | 'round';

export interface ButtonComponentProps {
	type?: ButtonType;
	size?: ButtonSize;
	icon?: ReactNode;
	iconPosition?: 'left' | 'right';
	loading?: boolean;
	loadingIcon?: ReactNode;
	loadingChildren?: ReactNode;
	disabled?: boolean;
	hairline?: boolean;
	plain?: boolean;
	shape?: ButtonShape;
	block?: boolean;
	color?: string;
	htmlType?: ButtonHTMLAttributes<HTMLButtonElement>['type'];
}

export type ButtonProps = ButtonComponentProps &
	Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'type'>;
