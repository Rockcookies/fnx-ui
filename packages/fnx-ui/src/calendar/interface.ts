import { HTMLAttributes, ReactNode } from 'react';
import CalendarDay from './CalendarDay';
import CalendarMulti from './CalendarMulti';
import CalendarRange from './CalendarRange';

export type CalendarValue = Date | string | number;

export type CalendarDayType =
	| 'normal'
	| 'start'
	| 'start-end'
	| 'middle'
	| 'end'
	| 'selected'
	| 'multiple-middle'
	| 'multiple-selected'
	| 'disabled';

export interface CalendarDayComponentProps {
	color?: string;
	dayHeight?: string;
	date: Date;
	content?: ReactNode;
	type: CalendarDayType;
	topInfo?: ReactNode;
	bottomInfo?: ReactNode;
}

export type CalendarDayProps = CalendarDayComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'content'>;

export interface CalendarSlots {
	day?: (day: CalendarDayComponentProps) => ReactNode;
	subTitle?: boolean | ((date: Date) => ReactNode);
	monthMark?: boolean | ((date: Date) => ReactNode);
	monthTitle?: boolean | ((date: Date) => ReactNode);
	footer?: ReactNode;
}

export interface CalendarBaseProps {
	title?: ReactNode;
	color?: string;
	minDate?: Date | string | number;
	maxDate?: Date | string | number;
	dayHeight?: number | string;
	readonly?: boolean;
	closeIcon?: ReactNode;
	firstDayOfWeek?: number;
	slots?: CalendarSlots;
	confirmText?: ReactNode;
	confirmDisabledText?: ReactNode;
}

export type CalendarElementProps = Omit<
	HTMLAttributes<HTMLDivElement>,
	'defaultValue' | 'onChange' | 'title'
>;

export interface CalendarContainerRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date[] | undefined;
}

export interface CalendarRef {
	element: HTMLDivElement | null;
	reset: () => void;
	getValue: () => Date | undefined;
}

export interface CalendarComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue;
	onChange?: (value: Date) => void;
	onConfirm?: (value?: Date) => void;
	onCancel?: () => void;
}

export type CalendarProps = CalendarComponentProps & CalendarElementProps;

export type CalendarRangeRef = CalendarContainerRef;

export interface CalendarRangeComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	allowSameDay?: boolean;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarRangeProps = CalendarRangeComponentProps &
	CalendarElementProps;

export type CalendarMultiRef = CalendarRangeRef;

export interface CalendarMultiComponentProps extends CalendarBaseProps {
	defaultValue?: CalendarValue[];
	onChange?: (value: Date[]) => void;
	onConfirm?: (value?: Date[]) => void;
	onCancel?: () => void;
	maxSize?: number;
	onMaxSize?: () => void;
}

export type CalendarMultiProps = CalendarMultiComponentProps &
	CalendarElementProps;

export interface CalendarExportExtra {
	Range: typeof CalendarRange;
	Multi: typeof CalendarMulti;
	Day: typeof CalendarDay;
}
