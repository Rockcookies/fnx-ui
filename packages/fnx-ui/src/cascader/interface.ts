import { HTMLAttributes, ReactNode } from 'react';

export type CascaderValue = string | number;

export interface CascaderOption {
	label?: ReactNode;
	value?: CascaderOption;
	loading?: boolean;
	disabled?: boolean;
	children?: CascaderOption[];
	// for custom filed names
	[key: string]: any;
}

export interface CascaderDataNames {
	label?: string | string[];
	value?: string | string[];
	loading?: string | string[];
	disabled?: string | string[];
	children?: string | string[];
}

export interface CascaderComponentProps<T = CascaderOption> {
	defaultValue?: CascaderValue[];
	data?: T[];
	dataNames?: CascaderDataNames;
	onLoadData?: (options: T[]) => void;
	onConfirm?: (value: CascaderValue[], options: T[]) => void;
	onCancel?: () => void;
	onChange?: (value: CascaderValue[], options: T[]) => void;
	title?: ReactNode;
	placeholder?: ReactNode;
	swipeable?: boolean;
	closeIcon?: ReactNode;
	activeColor?: string;
}

export type CascaderProps<T = CascaderOption> = CascaderComponentProps<T> &
	Omit<HTMLAttributes<HTMLDivElement>, 'defaultValue' | 'title' | 'onChange'>;
