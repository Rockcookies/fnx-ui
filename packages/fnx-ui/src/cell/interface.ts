import { HTMLAttributes, ReactNode } from 'react';
import CellGroup from './CellGroup';

export type CellItemsAlign = 'top' | 'middle' | 'bottom';

export interface CellComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	content?: ReactNode;
	contentProps?: HTMLAttributes<HTMLDivElement>;
	description?: ReactNode;
	descriptionProps?: HTMLAttributes<HTMLDivElement>;
	leftIcon?: ReactNode;
	leftIconProps?: HTMLAttributes<HTMLSpanElement>;
	rightIcon?: ReactNode;
	rightIconProps?: HTMLAttributes<HTMLSpanElement>;
	border?: boolean;
	itemsAlign?: CellItemsAlign;
	clickable?: boolean;
	size?: 'md' | 'lg';
}

export type CellProps = CellComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellGroupComponentProps {
	title?: ReactNode;
	titleProps?: HTMLAttributes<HTMLDivElement>;
	border?: boolean;
	insert?: boolean;
	bodyProps?: HTMLAttributes<HTMLDivElement>;
}

export type CellGroupProps = CellGroupComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'title'>;

export interface CellExportExtra {
	Group: typeof CellGroup;
}
