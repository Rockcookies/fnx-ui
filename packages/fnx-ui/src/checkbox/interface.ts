import { HTMLAttributes, ReactNode } from 'react';
import InternalCheckboxGroup from './CheckboxGroup';

export type CheckboxValue = string | number | boolean;

export interface CheckboxBaseProps {
	disabled?: boolean;
	direction?: 'vertical' | 'horizontal';
	checkedIcon?: ReactNode;
	icon?: ReactNode;
	iconSize?: string | number;
	iconPosition?: 'left' | 'right';
	iconShape?: 'square' | 'round' | 'plain';
	iconCheckedColor?: string;
	labelDisabled?: boolean;
}

export interface CheckboxComponentProps extends CheckboxBaseProps {
	value?: CheckboxValue;
	checked?: boolean;
	defaultChecked?: boolean;
	onChange?: (checked: boolean) => void;
	skipGroup?: boolean;
}

export type CheckboxProps = CheckboxComponentProps &
	Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'value'>;

export interface CheckboxRef {
	element: HTMLDivElement | null;
	toggle: () => void;
}

export interface CheckboxGroupComponentProps<
	T extends CheckboxValue = CheckboxValue,
> extends CheckboxBaseProps {
	value?: T[];
	defaultValue?: T[];
	onChange?: (v: T[]) => void;
	maxCheckedCount?: number;
}

export type CheckboxGroupProps<T extends CheckboxValue = CheckboxValue> =
	CheckboxGroupComponentProps<T> &
		Omit<HTMLAttributes<HTMLDivElement>, 'onChange' | 'defaultValue'>;

export interface CheckboxGroupContextData extends CheckboxBaseProps {
	value?: CheckboxValue[];
	registerValue?: (value: CheckboxValue) => void;
	cancelValue?: (value: CheckboxValue) => void;
	toggleValue?: (isCheck: boolean, value: CheckboxValue) => void;
}
export interface CheckboxGroupRef {
	element: HTMLDivElement | null;
	toggleAll: (options?: { checkAll?: boolean }) => void;
}

export interface CheckboxExportExtra {
	Group: typeof InternalCheckboxGroup;
}
