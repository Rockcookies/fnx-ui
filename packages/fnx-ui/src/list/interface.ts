import { HTMLAttributes, ReactNode } from 'react';

export type ListStatus = 'default' | 'loading' | 'finished' | 'error';

export type ListDirection = 'up' | 'down';

export interface ListComponentProps {
	status?: ListStatus;
	offset?: number;
	direction?: ListDirection;
	slots?: ListSlots;
	onLoad?: () => void;
	immediateCheck?: boolean;
	scrollListenTo?: HTMLElement | Window | (() => HTMLElement | Window) | null;
}

export type ListProps = ListComponentProps & HTMLAttributes<HTMLDivElement>;

export interface ListSlots {
	default?: ReactNode;
	loading?: ReactNode;
	finished?: ReactNode;
	error?: ReactNode;
}

export interface ListRef {
	root: HTMLDivElement | null;
	scrollParent: HTMLElement | Window | null;
	check: () => void;
}
