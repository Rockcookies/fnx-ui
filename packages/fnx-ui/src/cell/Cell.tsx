import React, { createElement, ReactNode } from 'react';
import { Dictionary } from '../utils/interface';
import { classnames, createBEM } from '../utils/namespace';
import { createDefaultsForwardRef } from '../utils/react';
import { CellProps } from './interface';

const NS = 'fnx-cell';
const bem = createBEM(NS);

const Cell = createDefaultsForwardRef<
	HTMLDivElement,
	CellProps,
	Required<
		Pick<
			CellProps,
			| 'border'
			| 'itemsAlign'
			| 'clickable'
			| 'titleProps'
			| 'descriptionProps'
			| 'contentProps'
			| 'leftIconProps'
			| 'rightIconProps'
		>
	>
>(
	'Cell',
	{
		border: true,
		itemsAlign: 'top',
		clickable: false,
		titleProps: {},
		descriptionProps: {},
		contentProps: {},
		leftIconProps: {},
		rightIconProps: {},
	},
	(
		{
			border,
			itemsAlign,
			clickable,
			titleProps,
			descriptionProps,
			contentProps,
			leftIconProps,
			rightIconProps,
			// optionals
			title,
			description,
			content,
			leftIcon,
			rightIcon,
			className,
			children,
			...restProps
		},
		ref,
	) => {
		const renderNode = (
			tag: string,
			className: string,
			props: Dictionary,
			node: ReactNode,
		): ReactNode => {
			if (node) {
				return createElement(
					tag,
					{
						...props,
						className: classnames(className, props.className),
					},
					node,
				);
			}
		};

		return (
			<div
				className={classnames(
					bem([
						`align-${itemsAlign}`,
						{ clickable, 'border-less': !border },
					]),
					className,
				)}
				role={clickable ? 'button' : undefined}
				tabIndex={clickable ? 0 : undefined}
				{...restProps}
				ref={ref}
			>
				{renderNode('span', bem('left-icon'), leftIconProps, leftIcon)}

				{(title || description) && (
					<div
						{...titleProps}
						className={classnames(
							bem('title'),
							titleProps.className,
						)}
					>
						{title}
						{renderNode(
							'div',
							bem('description'),
							descriptionProps,
							description,
						)}
					</div>
				)}

				{renderNode(
					'div',
					bem('content', { alone: !(title || description) }),
					contentProps,
					content || children,
				)}

				{renderNode(
					'span',
					bem('right-icon'),
					rightIconProps,
					rightIcon,
				)}
			</div>
		);
	},
);

export default Cell;
