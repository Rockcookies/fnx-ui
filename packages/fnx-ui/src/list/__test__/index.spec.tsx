import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import List from '../index';
import Cell from '../../cell';
import { ListSlots } from '../interface';

describe('<List/>', () => {
	const slots: ListSlots = {
		loading: 'loading',
		finished: 'finished',
		error: 'error',
	};

	beforeEach(() => {
		// mocking the offsetParent
		Object.defineProperty(HTMLElement.prototype, 'offsetParent', {
			get: function () {
				return {};
			},
		});
	});

	it('should emit load event when reaching bottom', () => {
		const onLoad = jest.fn();

		render(<List onLoad={onLoad} />);

		expect(onLoad).toHaveBeenCalled();
	});

	it('should render loading tip correctly and do not trigger onLoad', () => {
		const onLoad = jest.fn();

		const { container } = render(<List status="loading" onLoad={onLoad} />);

		expect(container).toMatchSnapshot();
		expect(onLoad).not.toBeCalled();
	});

	it('should render finish tip correctly and do not trigger onLoad', () => {
		const onLoad = jest.fn();

		const { container } = render(
			<List status="finished" onLoad={onLoad} />,
		);

		expect(container).toMatchSnapshot();
		expect(onLoad).not.toBeCalled();
	});

	it('should render error slots correctly and do not trigger onLoad', async () => {
		const onLoad = jest.fn();

		const { container } = render(<List slots={slots} status="error" />);

		expect(container).toMatchSnapshot();
		expect(onLoad).not.toBeCalled();
	});

	it('should render correctly when set direction', async () => {
		const onLoad = jest.fn();

		const { container, rerender } = render(
			<List slots={slots} direction="up" onLoad={onLoad} />,
		);

		rerender(
			<List
				slots={slots}
				direction="up"
				status="loading"
				onLoad={onLoad}
			/>,
		);

		rerender(
			<List
				slots={slots}
				direction="up"
				status="default"
				onLoad={onLoad}
			/>,
		);

		expect(container).toMatchSnapshot();
		expect(onLoad).toHaveBeenCalledTimes(2);
	});

	it('should trigger load when scroll', async () => {
		const onLoad = jest.fn();

		render(
			<List onLoad={onLoad} immediateCheck={false}>
				{new Array(50).fill(0).map((_, i) => (
					<Cell key={i} title={i} />
				))}
			</List>,
		);

		await waitFor(() => {
			fireEvent.scroll(window, { target: { pageYOffset: 10 } });
		});

		expect(onLoad).toHaveBeenCalledTimes(1);
	});
});
