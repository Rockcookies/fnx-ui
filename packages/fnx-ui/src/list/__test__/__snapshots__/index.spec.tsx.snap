// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List/> should render correctly when set direction 1`] = `
<div>
  <div
    aria-busy="false"
    class="fnx-list fnx-list--default"
    role="feed"
  >
    <div
      class="fnx-list__placeholder"
    />
  </div>
</div>
`;

exports[`<List/> should render error slots correctly and do not trigger onLoad 1`] = `
<div>
  <div
    aria-busy="false"
    class="fnx-list fnx-list--error"
    role="feed"
  >
    <div
      class="fnx-list__indicator"
    >
      error
    </div>
    <div
      class="fnx-list__placeholder"
    />
  </div>
</div>
`;

exports[`<List/> should render finish tip correctly and do not trigger onLoad 1`] = `
<div>
  <div
    aria-busy="false"
    class="fnx-list fnx-list--finished"
    role="feed"
  >
    <div
      class="fnx-list__placeholder"
    />
  </div>
</div>
`;

exports[`<List/> should render loading tip correctly and do not trigger onLoad 1`] = `
<div>
  <div
    aria-busy="true"
    class="fnx-list fnx-list--loading"
    role="feed"
  >
    <div
      class="fnx-list__indicator"
    >
      <span
        aria-busy="true"
        aria-live="polite"
        class="fnx-loading"
      >
        <span
          class="fnx-icon"
          style="font-size: 16px;"
        >
          <svg
            aria-hidden="true"
            class="fnx-icon__spinner"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="25 25 50 50"
            width="1em"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="20"
            />
          </svg>
        </span>
        <span
          class="fnx-loading__text"
        >
          Loading...
        </span>
      </span>
    </div>
    <div
      class="fnx-list__placeholder"
    />
  </div>
</div>
`;
